flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSinkITCase.java

```diff
@@ -0,0 +1,298 @@
+                                            getSinkRuntimeProvider(
+                                                    DynamicTableSink.Context context) {
+                                        return SinkProvider.of(
+                                                TestSink.newBuilder()
+                                                        .setWriter(new TestWriter(timestamps))
+                                                        .setCommittableSerializer(
+                                                                TestSink.StringCommittableSerializer
+                                                                        .INSTANCE)
+                                                        .build());
+                                    }
+                                })
+                        .build();
+        tableEnv.createTable("T1", sourceDescriptor);
+        assertPlan(tableEnv, "INSERT INTO T1 SELECT * FROM T1", false);
+    }
+
+    private static void addTimestamp(SharedReference<List<Long>> timestamps, Long timestamp) {
+        timestamps.applySync(l -> l.add(timestamp));
+    }
+
+    private static void assertPlan(
+            StreamTableEnvironment tableEnv,
+            String sql,
+            boolean containsStreamRecordTimestampInserter) {
+        Matcher<String> matcher = containsString("StreamRecordTimestampInserter(rowtime field: 2");
+        if (!containsStreamRecordTimestampInserter) {
+            matcher = not(matcher);
+        }
+        assertThat(tableEnv.explainSql(sql, ExplainDetail.JSON_EXECUTION_PLAN), matcher);
+    }
+
+    private static ResolvedSchema resolvedSchemaStreamRecordTimestampInserter() {
+        return ResolvedSchema.of(
+                Column.physical("a", INT()),
+                Column.physical("b", STRING()),
+                Column.physical("ts", TIMESTAMP_LTZ(3)));
+    }
+
+    private static Schema schemaStreamRecordTimestampInserter(
+            ResolvedSchema resolvedSchema, boolean withWatermark) {
+        Schema.Builder builder = Schema.newBuilder().fromResolvedSchema(resolvedSchema);
+        if (withWatermark) {
+            builder.watermark("ts", "ts");
+        }
+        return builder.build();
+    }
+
+    private static void assertTimestampResults(
+            SharedReference<List<Long>> timestamps, List<Row> rows) {
+        assertEquals(rows.size(), timestamps.get().size());
+        for (int i = 0; i < rows.size(); i++) {
+            assertEquals(rows.get(i).getField(2), Instant.ofEpochMilli(timestamps.get().get(i)));
+        }
+    }
+
+    private static class TimestampTestSource extends TableFactoryHarness.ScanSourceBase {
+
+        private final List<Row> rows;
+        private final DataType producedDataType;
+
+        private TimestampTestSource(List<Row> rows, DataType producedDataType) {
+            super(false);
+            this.producedDataType = producedDataType;
```

Comment:
I just remember that there is actually a way easier solution. We have prepared the test base already for retrieving the type. Take a look at the super class, you can access `getFactoryContext().getPhysicalDataType()`

Reply:
Thank you! fixed.